core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -437,41 +440,50 @@ class TimeSeriesShard(val dataset: Dataset,
   def ingest(data: SomeData): Long = ingest(data.records, data.offset)
 
   def recoverIndex(): Future[Unit] = {
-    val tracer = Kamon.buildSpan("memstore-recover-index-latency")
-      .withTag("dataset", dataset.name)
-      .withTag("shard", shardNum).start()
+    val p = Promise[Unit]()
+    Future {
+      assertThreadName(IngestSchedName)
+      val tracer = Kamon.buildSpan("memstore-recover-index-latency")
+        .withTag("dataset", dataset.name)
+        .withTag("shard", shardNum).start()
 
-    /* We need this map to track partKey->partId because lucene index cannot be looked up
+      /* We need this map to track partKey->partId because lucene index cannot be looked up
        using partKey efficiently, and more importantly, it is eventually consistent.
         The map and contents will be garbage collected after we are done with recovery */
-    val partIdMap = debox.Map.empty[BytesRef, Int]
-
-    val earliestTimeBucket = Math.max(0, currentIndexTimeBucket - numTimeBucketsToRetain)
-    logger.info(s"Recovering timebuckets $earliestTimeBucket to ${currentIndexTimeBucket - 1} " +
-      s"for dataset=${dataset.ref} shard=$shardNum ")
-    // go through the buckets in reverse order to first one wins and we need not rewrite
-    // entries in lucene
-    // no need to go into currentIndexTimeBucket since it is not present in cass
-    val timeBuckets = for { tb <- currentIndexTimeBucket-1 to earliestTimeBucket by -1 } yield {
-      colStore.getPartKeyTimeBucket(dataset, shardNum, tb).map { b =>
-        new IndexData(tb, b.segmentId, RecordContainer(b.segment.array()))
+      val partIdMap = debox.Map.empty[BytesRef, Int]
+
+      val earliestTimeBucket = Math.max(0, currentIndexTimeBucket - numTimeBucketsToRetain)
+      logger.info(s"Recovering timebuckets $earliestTimeBucket to ${currentIndexTimeBucket - 1} " +
+        s"for dataset=${dataset.ref} shard=$shardNum ")
+      // go through the buckets in reverse order to first one wins and we need not rewrite
+      // entries in lucene
+      // no need to go into currentIndexTimeBucket since it is not present in cass
+      val timeBuckets = for {tb <- currentIndexTimeBucket - 1 to earliestTimeBucket by -1} yield {
+        colStore.getPartKeyTimeBucket(dataset, shardNum, tb).map { b =>
+          new IndexData(tb, b.segmentId, RecordContainer(b.segment.array()))
+        }
       }
-    }
-    val fut = Observable.flatten(timeBuckets: _*)
-      .foreach(tb => extractTimeBucket(tb, partIdMap))(ingestSched)
-      .map(_ => completeIndexRecovery())
-    fut.onComplete(_ => tracer.finish())
-    fut
+      Observable.flatten(timeBuckets: _*)
```

Comment:
Is the change here the thread-safety bug you mention?

Reply:
Yes - thats right